### PR TITLE
Add heuristic for when `evict_first` should be set (and some other minor things)

### PR DESCRIPTION
**Pull request by Chillee**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108841

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @aakhundov @ColinPeppler @ngimel

Example of when the `evict_first` heuristic helps (although arguably we should be doing 2D tiling here).
```
@torch.compile
def f(a, b):
    return (a * b).sum(dim=-1)

N = 512
inps = (torch.randn(N, N, N).permute(2, 1, 0), torch.randn(N, N, N).permute(1, 2, 0))
from torch._inductor.utils import do_bench
print(do_bench(lambda: f(*inps)))
```

This generates code like this: http://ix.io/4HFs

```
Original: 3.8 ms
This PR: 3.54 ms
Always `evict_first: 5.4ms
```

### CODE REVIEW

File: torch/_inductor/utils.py [hunk 1]
```diff
@@ -310,7 +310,7 @@ def timed(model: Callable[..., Any], example_inputs, times: int = 1) -> float:
 
 def print_performance(fn, args=(), times=10, repeat=10, baseline=1.0):
     timings = torch.tensor([timed(fn, args, times) for _ in range(repeat)])
-    took = torch.median(timings)
+    took = torch.median(timings)/times
```
Comment (lezcano) [inline torch/_inductor/utils.py:313]: Mediand divided by times??

File: torch/_inductor/codegen/triton.py [hunk 2]
```diff
@@ -1314,7 +1314,8 @@ def load(self, name: str, index: sympy.Expr):
                 names = {name}
             last_use = len(names & self.last_usage) > 0
             evict_last = not last_use and ("rmask" in mask or indirect_indexing)
-            ep = ", eviction_policy='evict_last'" if evict_last else ""
+            ep = ", eviction_policy='evict_last'" if evict_last else ", eviction_policy='evict_first'"
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:1317]: For this, how do the benchmarks look like?

As a side note, I wanted to set the cache modifier to mark it as streaming data (`.cs`) but this is not implemented in triton yet, but I guess this is similar.

Comment (lezcano) [reply torch/_inductor/codegen/triton.py:1317]: Actually, adding `.cg` (which is supported by triton) for loads when there is no further use, and for stores when they are outside of a reduction loop should help with kernels that have spilled registers.

File: torch/_inductor/utils.py (hunk 1 above)
Comment (Chillee) [reply torch/_inductor/utils.py:313]: yeah, the benchmark runs the function 10 times and takes the overall runtime.

The median is computed over the "repeat" dimension, if that makes sense.

Comment (lezcano) [reply torch/_inductor/utils.py:313]: Ah, I see. Fair enough.

File: torch/_inductor/codegen/triton.py (hunk 2 above)
Comment (Chillee) [reply torch/_inductor/codegen/triton.py:1359]: @lezcano overall, there do seem to be some slight regressions.

I'd like to set `evict_first` *only* if there is another broadcasted load within the same loop. Any thoughts on that and the easiest way to accomplish it?

Comment (lezcano) [reply torch/_inductor/codegen/triton.py:1359]: In the scheduler you can find the loads in the current reduction looking at loads between EnableReduction and Disable reduction, and then you have the function `is_broadcast` to tell you whether any of those is going to broadcast, and in this function you have the conditions when a load is going to be lifted outside the reduction loop:
```python
        elif (
            self.inside_reduction
            and not self.persistent_reduction
            and "rmask" not in mask
            and not indirect_indexing
        ):
            # can lift a common load outside of reduction loop
            # One exception is when this is an indirect_load.
```

File: torch/_inductor/codegen/triton.py [hunk 3]
```diff
@@ -1341,6 +1341,28 @@ def _new_line(self, line):
 
         return sympy_symbol(str(var))
 
+    def get_strides_of_load(self, index: sympy.Expr):
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:1344]: any reason not to lift this to `common.py` and use it in triton and c++?

File: torch/_inductor/codegen/triton.py [hunk 4]
```diff
@@ -1353,16 +1375,23 @@ def load(self, name: str, index: sympy.Expr):
...
                 names = {name}
             last_use = len(names & self.last_usage) > 0
             evict_last = not last_use and ("rmask" in mask or indirect_indexing)
-            ep = ", eviction_policy='evict_last'" if evict_last else ""
+            if evict_last:
+                ep = ", eviction_policy='evict_last'"
+            else:
+                ep = ", eviction_policy='evict_first'"
         else:
             ep = ""
```
Comment (lezcano) [inline torch/_inductor/codegen/triton.py:1396]: evict_first ehre as well?

**Comment by Chillee** [comment]
image

Seems like it's about 1% improvement across the board.

File: torch/_inductor/codegen/triton.py (hunk 4 above)
Comment (Chillee) [reply torch/_inductor/codegen/triton.py:1396]: Possibly, but I think generally `evict_first` isn't something that should always be thrown in.

**Comment by lezcano** [comment]
image
image

There's two models that look like they may not be noise, everything else looks within noise range.

It's not regressing, so LGTM. The comments above are relevant and the `.cg` I think it'd be good. That being said, you'd need to regtriger a bench run for those

**Comment by Chillee** [comment]
@lezcano I think this change's results is in a large sense incremental improvements across the board, many of which are caught up in the noise threshold.

File: torch/_inductor/codegen/triton.py (hunk 3 above)
Comment (Chillee) [reply torch/_inductor/codegen/triton.py:1344]: Just don't have a particular need for it in C++ today.

File: torch/_inductor/codegen/triton.py (hunk 4 above)
Comment (lezcano) [reply torch/_inductor/codegen/triton.py:1396]: I think it makes sense to use it given how we structure our kernels, where most of the loads are at the beginning, and then we don't reload them again

File: torch/_inductor/codegen/triton.py (hunk 3 above)
Comment (lezcano) [reply torch/_inductor/codegen/triton.py:1344]: I meant
https://github.com/pytorch/pytorch/blob/f2d7faf4ba92a6ed43890775ca6ca174ddbf99ea/torch/_inductor/codegen/cpp.py#L293-L296

**Review by lezcano — APPROVED** [review]
Approving to unblock, as the benchmarks look good

**Review by jansel — APPROVED** [review]
